File: solution/1219. Path with Maximum Gold/1219.cpp
class Solution {
public:
    int getMaximumGold(vector<vector<int>>& grid) {
        m = grid.size(), n = grid[0].size();
        int res = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                res = max(res, dfs(grid, i, j));
            }
        }
        return res;
    }

private:
    const vector<int> dirs = {0, -1, 0, 1, 0};
    int m, n;

    int dfs(vector<vector<int>>& grid, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == 0) return 0;
        int val = grid[i][j];
        grid[i][j] = 0;
        int mx = 0;
        for (int k = 0; k < 4; ++k) {
            int y = i + dirs[k];
            int x = j + dirs[k + 1];
            mx = max(mx, dfs(grid, y, x));
        }
        grid[i][j] = val;
        return mx + val;
    }
};

File: solution/1219. Path with Maximum Gold/1219.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        DIR = [0, 1, 0, -1, 0]
        
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == 0:
                return 0
            val = grid[i][j]
            grid[i][j] = 0
            res = 0
            for k in range(4):
                res = max(res, dfs(i + DIR[k], j + DIR[k + 1]))
            grid[i][j] = val
            return res + val
        
        res = 0
        for i in range(m):
            for j in range(n):
                res = max(res, dfs(i, j))
        return res
    
